examples/step-78/step-78.cc

```diff
@@ -0,0 +1,869 @@
+  {
+    SolverControl                          solver_control(1000, 1e-12);
+    SolverCG<Vector<double>>               cg(solver_control);
+    PreconditionSSOR<SparseMatrix<double>> preconditioner;
+    preconditioner.initialize(system_matrix, 1.0);
+    cg.solve(system_matrix, solution, system_rhs, preconditioner);
+    constraints.distribute(solution);
+    std::cout << "     " << solver_control.last_step() << " CG iterations."
+              << std::endl;
+  }
+
+  /*
+  This is simply the function to build the solution together. For this, we
+  create a new layer at each time, and then add the solution vector for that
+  timestep. The function then stitches this together with the old solutions
+  using 'build_patches'.
+  */
+  template <int dim>
+  void BlackScholes<dim>::o_results()
+  {
+    data_out_stack.new_parameter_value(time, time_step);
+    data_out_stack.attach_dof_handler(dof_handler);
+    data_out_stack.add_data_vector(solution, solution_names);
+    data_out_stack.build_patches(2);
+    data_out_stack.finish_parameter_value();
+  }
+
+  /*
+  This is the function that refines the spatial mesh. This is exactly what was
+  used in Step-26, so there is nothing new here.
+  */
+  template <int dim>
+  void BlackScholes<dim>::refine_mesh(const unsigned int min_grid_level,
+                                      const unsigned int max_grid_level)
+  {
+    Vector<float> estimated_error_per_cell(triangulation.n_active_cells());
+    KellyErrorEstimator<dim>::estimate(
+      dof_handler,
+      QGauss<dim - 1>(fe.degree + 1),
+      std::map<types::boundary_id, const Function<dim> *>(),
+      solution,
+      estimated_error_per_cell);
+    GridRefinement::refine_and_coarsen_fixed_fraction(triangulation,
+                                                      estimated_error_per_cell,
+                                                      0.6,
+                                                      0.4);
+    if (triangulation.n_levels() > max_grid_level)
+      for (const auto &cell :
+           triangulation.active_cell_iterators_on_level(max_grid_level))
+        cell->clear_refine_flag();
+    for (const auto &cell :
+         triangulation.active_cell_iterators_on_level(min_grid_level))
+      cell->clear_coarsen_flag();
+    SolutionTransfer<dim> solution_trans(dof_handler);
+    Vector<double>        previous_solution;
+    previous_solution = solution;
+    triangulation.prepare_coarsening_and_refinement();
+    solution_trans.prepare_for_coarsening_and_refinement(previous_solution);
+    triangulation.execute_coarsening_and_refinement();
+    setup_system();
+    solution_trans.interpolate(previous_solution, solution);
+    constraints.distribute(solution);
+  }
+
+  /*
+  This is somewhat unnecessary to have a function for the global refinement that
+  we do. The reason for the function, is to allow for the possibility of an
+  adaptive refinement later.
+  */
+
+  template <int dim>
```

Comment:
```suggestion
```

Reply:
I'm not sure what you're saying with this. You don't want me to remove these lines do you?